crates/bevy_ecs/src/storage/mod.rs

```diff
@@ -41,3 +43,71 @@ pub struct Storages {
     /// Backing storage for `!Send` resources.
     pub non_send_resources: Resources<false>,
 }
+
+/// Provides interior-mutable access to a world's internal data storages.
+///
+/// Any instance of this type is associated with a set of world data that
+/// it is allowed to access. This should be described in the documentation
+/// of wherever you obtained the `UnsafeStorages`.
+///
+/// For instance, if you originally obtained it from a system running on
+/// a multi-threaded executor, then you are only allowed to access data
+/// that has been registered in the system's `archetype_component_access`.
+/// If you originally obtained an `UnsafeStorages` from an `&World`,
+/// then you have read-only access to the entire world.
+///
+/// Accessing world data that do not have access to, or mutably accessing
+/// data that you only have read-access to, is considered undefined behavior.
+pub struct UnsafeStorages<'a> {
+    pub sparse_sets: UnsafeSparseSets<'a>,
+    pub tables: UnsafeTables<'a>,
+    pub resources: UnsafeResources<'a, true>,
+    pub non_send_resources: UnsafeResources<'a, false>,
+}
+
+impl<'a> UnsafeStorages<'a> {
+    pub(crate) fn new(storages: &'a Storages) -> Self {
+        Self {
+            sparse_sets: UnsafeSparseSets {
+                sparse_sets: &storages.sparse_sets,
+            },
+            tables: UnsafeTables {
+                tables: &storages.tables,
+            },
+            resources: UnsafeResources {
+                resources: &storages.resources,
+            },
+            non_send_resources: UnsafeResources {
+                resources: &storages.non_send_resources,
+            },
+        }
+    }
+}
+
+/// A view into the [`ComponentSparseSet`] collection of [`UnsafeStorages`].
+#[derive(Clone, Copy)]
+pub struct UnsafeSparseSets<'a> {
+    sparse_sets: &'a SparseSets,
+}
+
+impl<'a> UnsafeSparseSets<'a> {
+    /// Gets a view into the [`ComponentSparseSet`] associated with `component_id`.
+    pub fn get(self, component_id: ComponentId) -> Option<UnsafeComponentSparseSet<'a>> {
```

Comment:
Perhaps I'm missing something obvious: why do these methods accept `self` and return an owned object, rather than working by reference?

Reply:
These types are thin wrappers around a single reference and implement `Copy`, but are still bound by the lifetime of the reference. Essentially, using `self` *is* working by reference. This is the same way that `UnsafeWorldCell` works.

Reply:
Ah, thanks for the explanation! That makes much more sense. Perhaps a note in the docs that these serve as smart pointer types would help?

Reply:
I don't think it's really accurate to call it a smart pointer. It's just a `Copy` type that stores a reference.

Currently these types are described as a "view" into world storage. I think that does a good enough job at describing that these types work by reference, but let me know if you think I could make it clearer somehow.